File: core/SpiralMatrix/SpiralMatrix.java
package com.dbc;

import java.util.ArrayList;
import java.util.List;

public class SpiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int dir = 0, row = 0, col = 0;
        int rows = matrix.length, cols = matrix[0].length;
        int[][] dirStep = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        while (row >= 0 && row < rows && col >= 0 && col < cols && matrix[row][col] != 1000) {
            res.add(matrix[row][col]);
            matrix[row][col] = 1000;
            int nextRow = row + dirStep[dir][0], nextCol = col + dirStep[dir][1];
            if (!(nextRow >= 0 && nextRow < rows && nextCol >= 0 && nextCol < cols && matrix[nextRow][nextCol] != 1000)) {
                dir = (dir + 1) % 4;
            }
            row += dirStep[dir][0];
            col += dirStep[dir][1];
        }
        return res;
    }
}

File: core/SpiralMatrix/SpiralMatrix.py
from typing import List

def spiralOrder(matrix: List[List[int]]) -> List[int]:
    res, direct, dir_step = [], 0, [[0, 1], [1, 0], [0, -1], [-1, 0]]
    rows, cols, row, col = len(matrix), len(matrix[0]), 0, 0
    while 0 <= row < rows and 0 <= col < cols and matrix[row][col] != 1000:
        res.append(matrix[row][col])
        matrix[row][col] = 1000
        next_row, next_col = row + dir_step[direct][0], col + dir_step[direct][1]
        if not (0 <= next_row < rows and 0 <= next_col < cols and matrix[next_row][next_col] != 1000):
            direct = (direct + 1) % 4

        row += dir_step[direct][0]
        col += dir_step[direct][1]

    return res



